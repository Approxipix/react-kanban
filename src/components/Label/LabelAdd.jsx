import React, { Component } from "react";
import PropTypes from 'prop-types';
import { connect } from "react-redux";
import { bindActionCreators } from 'redux';
import { addLabelToCard } from "../../redux/cardReducer/actions/actions";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import styled from 'styled-components';
import Label from "./Label";

const LabelList = styled.ul`
  display: flex;
  flex-direction: column;
  padding: 0 .5rem;
`;

const LabelItem = styled.li`
  position: relative;
  height: 2.25rem;
  width: 100%;
  margin-bottom: .3rem;
  background-color: ${props => props.value};
  cursor: pointer;
  border-radius: .2rem;
  &:before {
    content: '';
    position: absolute;
    top: 0; 
    bottom: 0; 
    left: 0;
    width: .6rem;
    border-radius: .2rem 0 0 .2rem;
    background-color: ${props => props.value};
    box-shadow: inset 2rem 0 rgba(0, 0, 0, .2);
    transition: transform .2s ease-in-out;
    z-index: -1;
  }
  &:hover {
    &:before {
      transform: translateX(-90%);
    }
  }
`;

const LabelIcon = styled.div`
  color: #fff;
  position: absolute;
  top: 50%;
  left: 50%;
  font-size: 1rem;
  transform: translate(-50%, -50%);
`;

class LabelAdd extends Component {
  addLabel = (label) => {
    this.props.actions.addLabelToCard({
      cardId: this.props.cardId,
      cardLabel: label
    })
  };

  render() {
    const { cardLabels, labels } = this.props;
    if (!labels) return null;
    return (
      <LabelList>
        {labels.map((label, index) => (
          <LabelItem
            key={index}
            value={label.color}
            onClick={() => this.addLabel(label)}
          >
            {cardLabels && cardLabels.some(cardLabel => cardLabel._labelId === label._labelId) && (
              <LabelIcon>
                <FontAwesomeIcon icon="check" />
              </LabelIcon>
            )}
          </LabelItem>
        ))}
      </LabelList>
    )
  };
}

LabelAdd.propTypes = {
  labels: PropTypes.arrayOf(
    PropTypes.shape({
      _labelId: PropTypes.number.isRequired,
      color: PropTypes.string.isRequired
    })
  ),
  cardLabels: PropTypes.arrayOf(
    PropTypes.shape({
      _labelId: PropTypes.number.isRequired,
      color: PropTypes.string
    })
  ),
};

function mapStateToProps(state) {
  return {
    labels: state.rootReducer.labels,
  }
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      addLabelToCard: addLabelToCard,
    }, dispatch)
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(LabelAdd);
