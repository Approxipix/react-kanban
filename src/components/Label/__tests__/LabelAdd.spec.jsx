import React from 'react'
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme'
import toJson from "enzyme-to-json";
import LabelAdd from '../LabelAdd'

const mockStore = configureStore([]);

describe('<LabelAdd>', () => {
  let store;
  let wrapper;
  let component;

  beforeEach(() => {
    store = mockStore({
      rootReducer: {
        labels: [
          { _labelId: 0, color: '#61BC4F' },
          { _labelId: 1, color: '#F2D600' },
          { _labelId: 2, color: '#FF9F1A' },
        ]
      }
    });
    store.dispatch = jest.fn();

    wrapper = shallow(<LabelAdd store={store}  />);
    component = wrapper.find('LabelAdd').dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render a label item for each labels from store', () => {
    expect(component.children().length).toEqual(3)
  });

  it('should sets the color prop as the `value` on the Label item', () => {
    expect(component.childAt(0).props().value).toEqual('#61BC4F');
    expect(component.childAt(1).props().value).toEqual('#F2D600');
    expect(component.childAt(2).props().value).toEqual('#FF9F1A');
  });

  it('should renders fontawesome icon for label with same labelID', () => {
    const props = { cardLabels: [{_labelId: 1}] };
    const wrapper = shallow(<LabelAdd store={store} {...props} />);
    const component = wrapper.find('LabelAdd').dive();

    expect(toJson(component)).toMatchSnapshot()
  });

  it('should dispatch an action on Label Item click', () => {
    const props = { cardId: 'CardID' };
    const wrapper = shallow(<LabelAdd store={store} {...props} />);
    const component = wrapper.find('LabelAdd').dive();
    const expectedAction = {
      type: 'ADD_LABEL_TO_CARD',
      payload: {
        cardId: 'CardID',
        cardLabel: { _labelId: 0, color: '#61BC4F' }
      }
    };

    component.childAt(0).simulate('click');

    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(expectedAction);
  });
});
