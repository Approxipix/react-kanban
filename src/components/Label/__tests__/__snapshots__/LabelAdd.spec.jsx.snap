// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LabelAdd> should render with given state from Redux store 1`] = `
<styled.ul>
  <styled.li
    key="0"
    onClick={[Function]}
    value="#61BC4F"
  />
  <styled.li
    key="1"
    onClick={[Function]}
    value="#F2D600"
  />
  <styled.li
    key="2"
    onClick={[Function]}
    value="#FF9F1A"
  />
</styled.ul>
`;

exports[`<LabelAdd> should renders fontawesome icon for label with same labelID 1`] = `
<styled.ul>
  <styled.li
    key="0"
    onClick={[Function]}
    value="#61BC4F"
  />
  <styled.li
    key="1"
    onClick={[Function]}
    value="#F2D600"
  >
    <styled.div>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="check"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </styled.div>
  </styled.li>
  <styled.li
    key="2"
    onClick={[Function]}
    value="#FF9F1A"
  />
</styled.ul>
`;
